[문승준] Mission 3: 회원가입 기능 구현 1차 수정

palza4dev opened the pull request:
- user 앱 views에 SignUpView class 추가
- 예외처리 (이메일 및 비밀번호 공란, 키에러, 이메일 중복)
- 이메일, 비밀번호, 생년월일 양식은 정규표현식으로 validate
- models에 적용 안된 부분 발견하여 마이그레이션 진행
- _Validation을 위한 정규표현식 코드 간소화 (추가)_

>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- user 앱 views에 회원가입 기능을 위한 SignUpView class 작성
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 이메일이나 패스워드가 전달되지 않으면 Key Error 반환
- 이메일이나 패스워드가 빈문자열이면 400 에러 반환
- 이메일 중복시 400 에러 반환
- 이메일, 패스워드, 생년월일 정규표현식에 맞지 않으면 400 에러 반환

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- models의 date_birth 필드 옵션을 null=True로 했으나, 회원가입시 null 입력이 가능하도록 구현하지 못함.
- 이전에 수정한 models 내용을 migrate 해놓지 않아서 다급하게 진행함. -> 다음부턴 주의 필요

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

## 회원가입 테스트와 DB 테이블 캡쳐화면
![스크린샷 2021-09-27 오후 5 11 47](https://user-images.githubusercontent.com/72376931/134872450-5fb1f9fc-3de6-418b-888c-8fb8c0db5c69.png)

palza4dev committed:
<Add> user 앱 회원가입 기능 추가
- user 앱 views에 SignUpView class 추가
- 예외처리 (이메일 및 비밀번호 공란, 키에러, 이메일 중복)
- 이메일, 비밀번호, 생년월일 양식은 정규표현식으로 validate
- models에 적용 안된 부분 발견하여 마이그레이션 진행

lordmyshepherd commented on students/seungjun/users/views.py:
```diff
@@ -1,3 +1,51 @@
+            if User.objects.filter(email=email).exists():
+                return JsonResponse({"message":"Email already exists"}, status=400)
+
+            regex_email = re.match(
+                '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$', email)
+            if not regex_email:
+                return JsonResponse({'message':'Email format is not valid'}, status=400)
+
+            regex_password = re.match(
+                '^(?=.*[A-Za-z])(?=.*\d)(?=.*[?!@#$%*&])[A-Za-z\d?!@#$%*&]{8,}$', password)
+            if not regex_password:
+                return JsonResponse({'message':'Password format is not valid'}, status=400)
+
+            regex_date_birth = re.match(
+                '^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])$', date_birth)
+            if not regex_date_birth: 
+                return JsonResponse({'message':'Date format must be in YYYY-MM-DD'}, status=400)
```
불필요하게 match() 함수 호출 후에 반환되는 값을 변수에 할당하지 않고 바로 if 문으로 True, False를 판별하도록 합니다. 읽어야할 코드의 양을 줄이고 가독성을 높이는 것은 굉장히 중요합니다.

```python
class SignUpView(View):
    def post(self, request):
        try:
            data = json.loads(request.body)
            name         = data['name']
            ..

            if not re.match('^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$', email)
                return JsonResponse({'message':'Email format is not valid'}, status=400)

            if not re.match('^(?=.*[A-Za-z])(?=.*\d)(?=.*[?!@#$%*&])[A-Za-z\d?!@#$%*&]{8,}$', password):
                return JsonResponse({'message':'Password format is not valid'}, status=400)

            if not re.match('^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])$', date_birth): 
                return JsonResponse({'message':'Date format must be in YYYY-MM-DD'}, status=400)
```

palza4dev replied:
네 수정하겠습니다.

palza4dev committed:
<Style> 더 나은 가독성을 위한 정규 표현식 코드 수정
- 이메일, 비밀번호, 생년월일 validation 코드 포맷팅 수정
- 불필요한 변수 할당 제거
- 코드의 양을 줄이고 가독성을 높이기 위함

kyeongraekim commented on students/seungjun/users/views.py:
```diff
@@ -1,3 +1,45 @@
+    def post(self, request):
+        data = json.loads(request.body)
+        try:
+            name         = data['name']
+            email        = data['email']
+            password     = data['password']
+            phone_number = data['phone_number']
+            gender       = data['gender']
+            date_birth   = data['date_birth']
```
현재 작성하신 데이터의 키 값을 끌어오는 코드는 null 값에 대한 처리가 없이 모두 키에러가 날 가능성이 있습니다.
data.get() 메소드의 사용법을 익혀보시고 null 값이 허용되어도 되는 추가정보사항에 적용시켜보세요~

palza4dev replied:
null 값이 허용되는 date_birth를 수정하고 validation TypeError도 처리하겠습니다.

kyeongraekim commented on students/seungjun/users/views.py:
```diff
@@ -1,3 +1,45 @@
+            date_birth   = data['date_birth']
+
+            if password == '' or email == '':
+                return JsonResponse({"message":"Email or Password cannot be empty"}, status=400)
```
현재 모델링과 뷰 기획상  password와 email 이 빈 스트링 값으로 들어올 수 있을까요?
해당 조건문의 활용성 및 필요성 여부를 판단해서 수정해주세요~

palza4dev replied:
정규 표현식으로 Validation 가능하기 때문에 필요없는 조건문으로 판단됩니다. 삭제하겠습니다!

kyeongraekim reviewed (changes requested):
리뷰 내용 확인해보시고 수정해주세요~

kyeongraekim commented on students/seungjun/westagram/urls.py:
```diff
@@ -13,7 +13,8 @@
 
 urlpatterns = [
-]
+    path('users/', include('users.urls'))
```
westargram/urls.py 가장 기본 path 디렉토리 에서 부터 slash를 추가 시키는 방법 보다, users/urls.py 디렉토리 앞선에 slash를 추가하는 방법을 적용해보세요~

palza4dev replied:
프로젝트의 path가 아닌 각 app의 path route 앞쪽에 slash 를 추가하겠습니다.

palza4dev committed:
<Modify> SignUpView date_birth 일부 수정 및 url path 수정
- date_birth은 Null값이 가능해서 get() 메소드를 이용해 키에러를 방지
- date_birth가 None인 경우 re.match() 함수 타입에러 발생 ->  None이 아닌
  경우에만 Validate하도록 if문 추가
- email, password가 빈문자열인 경우 정규표현식으로 걸러지므로 if문
  에러처리 삭제
- url  path의  slash는 app의  url route 앞에만  붙여주도록 수정함

lordmyshepherd commented:
LGTM 👍
